Feature: Early verification of account modifications in `BorrowedAccount`

Lichtso opened the pull request:
#### Problem
This is a continuation of #25380.

Currently, for BPF programs the verification cost is:
- One data copy from `AccountSharedData` to create or update the `PreAccount`
- One data copy from the BPF program into the `AccountSharedData` so we can compare it against `PreAccount`
- The actual comparison in `PreAccount` (this is skipped if the data is allowed to be modified)

Assuming that copies and comparison incur a similar cost (as they both need to access two different pointers, polluting the data cache), we could reduce this cost by 50% up to 66% for every account in every ABIv0 and ABIv1 BPF program executed. Built-in programs would also profit, reducing their verification cost by almost 100%. In ABIv2 BPF programs will be like built-in programs in this regard.

#### Summary of Changes
This PR introduces the feature `enable_early_verification_of_account_modifications` which controls the switch from the current behavior to the following:

1. Account modification errors are thrown earlier (hence the name of the feature). Specifically, they occur the moment they are attempted, not at the begin / end of instructions and CPI. The primary motivation is to get rid of `PreAccount` and all the redundant data copies and data comparisons that come with it. As a side effect, this should make debugging easier because we can get the call stack where the error occurs.

2. Similarly to the deprecation of `PreAccount`, `AccountsDataMeter` will also be deprecated and its functionality replicated inside `TransactionContext`. This was necessary as the `AccountsDataMeter` is currently updated and verified from withing the `PreAccount`s.

3. As a consequence, first corrupting (meta)data and then later correcting it into a valid state again won't fly anymore. Every change will have to be valid on its own. This however, does only apply to the runtime and built-in programs, not to BPF programs of ABIv0 and ABIv1 (all currently deployed programs) as they will continue to have their checks at the deserialization edge.

4. Failed attempted account modifications used to be counted as modifications, now they are not recorded as "touched" anymore.

5. All the explicit / dedicated account modification tests inside `InvokeContext`, `PreAccount` and `AccountsDataMeter` are disabled / become unreachable.

6. The loader and built-in tests on the other hand, become stricter as they now have to adhere to all account modification rules.

7. Two tests to trigger `InstructionError::UnbalancedInstruction` (which was not covered before) were added.

---

Feature Gate Issue: #26589

Lichtso commented on sdk/src/transaction_context.rs:
```diff
@@ -657,6 +720,38 @@ impl<'a> BorrowedAccount<'a> {
+    pub fn is_data_mutable(&self, new_length: usize) -> Result<(), InstructionError> {
+        if !self.transaction_context.enable_early_access_violation {
+            return Ok(());
+        }
```
https://github.com/solana-labs/solana/blob/98176fad773417a73dbd5bb4f08b9619f049d52e/program-runtime/src/pre_account.rs#L100

Lichtso commented on sdk/src/transaction_context.rs:
```diff
@@ -635,6 +676,28 @@ impl<'a> BorrowedAccount<'a> {
 
     /// Configures whether this account is executable (transaction wide)
     pub fn set_executable(&mut self, is_executable: bool) -> Result<(), InstructionError> {
+        if self.transaction_context.enable_early_access_violation {
```
https://github.com/solana-labs/solana/blob/98176fad773417a73dbd5bb4f08b9619f049d52e/program-runtime/src/pre_account.rs#L117

Lichtso commented on programs/bpf_loader/src/serialization.rs:
```diff
@@ -169,18 +169,25 @@ pub fn deserialize_parameters_unaligned(
+            // The redundant `is_data_mutable` helps to avoid the expensive data comparison if we can
+            if borrowed_account.is_data_mutable(data.len()).is_ok()
+                || borrowed_account.get_data() != data
+            {
```
As long as `enable_early_access_violation` is `false`, `is_data_mutable()` always returns `Ok(())` and this here evaluates to `if true`. So the behavior stays the same as it currently is.

Lichtso commented on programs/bpf_loader/src/serialization.rs:
```diff
@@ -344,14 +352,23 @@ pub fn deserialize_parameters_aligned(
+            // The redundant `is_data_mutable` helps to avoid the expensive data comparison if we can
+            if borrowed_account.is_data_mutable(data.len()).is_ok()
+                || borrowed_account.get_data() != data
+            {
```
As long as `enable_early_access_violation` is `false`, `is_data_mutable()` always returns `Ok(())` and this here evaluates to `if true`. So the behavior stays the same as it currently is.

Lichtso commented on sdk/src/transaction_context.rs:
```diff
@@ -682,3 +777,15 @@ impl From<TransactionContext> for ExecutionRecord {
         }
     }
 }
+
```
https://github.com/solana-labs/solana/blob/98176fad773417a73dbd5bb4f08b9619f049d52e/program-runtime/src/pre_account.rs#L180

jackcmay commented:
I'm not too worried about rent_epoch anymore since accounts must be rent-exempt.  We will either have to continue setting it for ABIv1, or maybe feature gate a deprecation of it which sets it to some acceptable value that is least likely to break existing programs.

jackcmay commented on programs/bpf_loader/src/serialization.rs:
```diff
@@ -167,18 +167,25 @@ pub fn deserialize_parameters_unaligned(
+                .get(start..start + pre_len)
+                .ok_or(InstructionError::InvalidArgument)?;
+            // The redundant `is_data_mutable` helps to avoid the expensive data comparison if we can
+            if borrowed_account.is_data_mutable(data.len()).is_ok()
```
When `enable_early_verification_of_account_modifications` is true, does this mean that if the account is not mutable we don't check the data?  Is that because if not mutable the program fails immediately if it attempts to write to it?

Lichtso replied:
For the runtime and built-in programs that would be the case yes, but this here is the deserialization edge (same goes for the CPI syscall) of BPF programs with ABIv0 or ABIv1. And once `enable_early_verification_of_account_modifications` is active, the following will be the case:
- BPF programs with ABIv0 or ABIv1 will continue to experience their errors at the Instruction or CPI edges, but here in the deserialization and the CPI syscall, instead of in `PreAccount::verify()`. So it is still "earlier" in that sense.
- Accounts for which `is_data_mutable() == false` will check if their data changed, and if so try to attempt to modify it, which will result in the correct Error being thrown. That will cost one data comparison, instead of the copy and comparison now.
- Accounts for which `is_data_mutable() == true` will not check if their data changed, and directly proceed to modify / overwrite it. That will cost one data copy, instead of the copy and comparison now.

Lichtso commented on program-test/src/lib.rs:
```diff
@@ -174,16 +174,27 @@ pub fn builtin_process_instruction(
+        {
             borrowed_account.set_data(&data)?;
         }
+        // if borrowed_account.is_executable() != is_executable {
+        //     borrowed_account.set_executable(is_executable)?;
+        // }
```
This is just a reminder that we don't support setting the `is_executable` flag here. Should probably be removed.

ryoqun commented on sdk/src/transaction_context.rs:
```diff
@@ -682,19 +810,101 @@ impl<'a> BorrowedAccount<'a> {
+        let mut changed_account_count = 0u64;
+        let mut total_size_of_all_accounts = 0u64;
+        let mut total_size_of_touched_accounts = 0u64;
+        for (index_in_transaction, flag) in
```
nit: `flag` => `touched` or `is_touched`. (slight readability improvement)

ryoqun commented on sdk/src/transaction_context.rs:
```diff
@@ -65,16 +70,20 @@ impl TransactionContext {
             return_data: TransactionReturnData::default(),
             total_resize_limit,
             total_resize_delta: RefCell::new(0),
+            enable_early_verification_of_account_modifications: false,
+            rent: Rent::default(),
```
I'd like to see these new `::default()` initializations go away if possible. I'm a bit concerned that we accidentally introduce a code path where `enable_early_verification_of_account_modifications()` won't be called, resulting in wrong or missed runtime checks. obviously, it's not currently not existing.

seems like changing the signature of `TransactionContext::new()` isn't not that hard, right? you can just explicitly pass defaults to reduce plumbing from non-critical code pathes like bench.

ryoqun commented on sdk/src/transaction_context.rs:
```diff
@@ -631,7 +733,7 @@ impl<'a> BorrowedAccount<'a> {
 
     /// Serializes a state into the account data
     pub fn set_state<T: serde::Serialize>(&mut self, state: &T) -> Result<(), InstructionError> {
-        let data = self.account.data_as_mut_slice();
```
nice catch :)

ryoqun commented on runtime/src/bank.rs:
```diff
@@ -4375,9 +4376,31 @@ impl Bank {
+            total_resize_delta,
         } = transaction_context.into();
         loaded_transaction.accounts = accounts;
+        if self
+            .feature_set
+            .is_active(&enable_early_verification_of_account_modifications::id())
+        {
+            saturating_add_assign!(
+                timings.details.total_account_count,
+                loaded_transaction.accounts.len() as u64
+            );
+            saturating_add_assign!(timings.details.changed_account_count, changed_account_count);
+            saturating_add_assign!(
+                timings.details.total_data_size,
+                total_size_of_all_accounts as usize
+            );
+            saturating_add_assign!(
+                timings.details.data_size_changed,
+                total_size_of_touched_accounts as usize
+            );
```
here

ryoqun commented on sdk/src/transaction_context.rs:
```diff
@@ -555,6 +611,28 @@ impl<'a> BorrowedAccount<'a> {
+                return Err(InstructionError::ModifiedProgramId);
+            }
+        }
+        self.touch()?;
```
these `.touch()` invocations are really need to be outside the feature flag? I think `self.touch()` is almost all cases infallible currently. but,  if they're not actually (in the future), this might risk cluster wide fork while the cluster is mixed with this new or original code, right?

it looks like the ultimate use is actually feature-gated here: https://github.com/solana-labs/solana/pull/25899/files#r916566935 (via not-feature gated new summation code in the `From`; this might be moved inside a feature gate maybe?)

ryoqun replied:
also, I'm wondering where failed account mutations should be counted towards to the summation or not. maybe i need precise semantics clarification (still digging relevant code).

i mean, data mutations which is prevented by the new checks won't be marked as `touch`-ed. So, failed ExecutionRecord in such a way won't include these attempts.

jstarry replied:
@Lichtso I agree with Ryo's points here, can you take another look at this in case you lost track of these review comments?

Lichtso replied:
@ryoqun @jstarry We could move the `touch()` inside a `if new_value != old_value` up to the beginning of each setter. That way failed attempts would also count as touched.

However, the statistic would still be different. Because currently one can fail many attempts inside one Instruction, but after this feature the first failed attempt is always also the last one. So it implicitly limits the number of possible failures to one at maximum.

jstarry replied:
I was agreeing with Ryo's first suggestion of moving `touch()` inside the feature gated code block. Specifically after the equality check.

I just looked closer into failed account mutations and agree that some of the stats shouldn't be updated (changed accounts and changed data size) but I think it's best to try to maintain the current behavior for now and address that later.

ryoqun commented on sdk/src/transaction_context.rs:
```diff
@@ -615,9 +715,11 @@ impl<'a> BorrowedAccount<'a> {
         let mut total_resize_delta = self.transaction_context.total_resize_delta.borrow_mut();
         *total_resize_delta = total_resize_delta
             .saturating_add((new_length as i64).saturating_sub(self.get_data().len() as i64));
+        self.touch()?;
```
why this `.touch()` isn't immediately following the `is_data_mutable()` calls like all other call-sites?

I think consistency is kind of necessity rather than nicety, considering the criticality of these code. :)

jstarry commented on sdk/src/transaction_context.rs:
```diff
@@ -682,19 +810,101 @@ impl<'a> BorrowedAccount<'a> {
+                .total_resize_delta
+                .borrow()
+                .saturating_add((new_length as i64).saturating_sub(self.get_data().len() as i64));
+            if total_resize_delta > self.transaction_context.total_resize_limit as i64 {
```
This check should be behind the `cap_accounts_data_len` feature

Lichtso replied:
@brooksprumo ATM, if this feature were to be enabled, the account data allocation check inside the transaction processing would be gone. See the second last commit "Relaxes AccountDataMeter related test cases". So, I think there should be some replacement in another PR, and the feature gate issue needs to keep track of it.

Lichtso commented on programs/bpf/tests/programs.rs:
```diff
@@ -305,16 +305,29 @@ fn run_program(name: &str) -> u64 {
                 }
             }
-            deserialize_parameters(
+            assert!(match deserialize_parameters(
```
Removing it would mean to remove the `deserialize_parameters` all together in this test, which is possible. The match is necessary as some errors (those coming from `enable_early_verification_of_account_modifications`) are expected here.

jstarry replied:
Yeah I think it seems ok to remove. @jackcmay do you think we could miss important issues in bpf program tests if remove this?

ryoqun commented on sdk/src/transaction_context.rs:
```diff
@@ -682,19 +810,101 @@ impl<'a> BorrowedAccount<'a> {
+            let account_data_size = context
+                .get_account_at_index(index_in_transaction)
+                .map(|account| account.borrow().data().len())
+                .unwrap_or(0) as u64;
```
nit: i think the `_or(0)` part should never hit in production, unless something very smelly is ongogigng, right? in that case, I'd recommend the hard `.expect("no out-of-range index_in_transaction")` or similar, because we're already introducing a new panic source in `.account_touched_flags.{borrow,borrow_mut}()`

ryoqun commented on programs/bpf_loader/src/syscalls.rs:
```diff
@@ -2672,26 +2673,50 @@ where
+                        .map_err(SyscallError::InstructionError)?;
+                }
+                if callee_account.get_owner() != caller_account.owner {
+                    callee_account
```
nit: how about commenting this very nuanced last placement of `.owner` like the other places?: 

```rust
// Change the owner at the end so that we are allowed to change the lamports and data before
```

ryoqun commented on sdk/src/transaction_context.rs:
```diff
@@ -706,7 +916,22 @@ impl From<TransactionContext> for ExecutionRecord {
+
+fn is_zeroed(buf: &[u8]) -> bool {
+    const ZEROS_LEN: usize = 1024;
+    static ZEROS: [u8; ZEROS_LEN] = [0; ZEROS_LEN];
```
nit: you can now use `const` here. :) this carried code is surely dated from pre const rusts. ;)

Lichtso commented:
@ryoqun @jstarry Can you verify that all your suggestions were implemented correctly?

Also, @jackcmay There are tow open question for you, one [here](https://github.com/solana-labs/solana/pull/25899#discussion_r916624449) the other is:

Can we get rid of this error kind (`AccountDataSizeChanged`):
```Rust
        // Only the owner can change the length of the data
        if !self.is_owned_by_current_program() && new_length != self.get_data().len() {
            return Err(InstructionError::AccountDataSizeChanged);
        }
```

As resize always implies a data change, which is already covered by (`ExternalAccountDataModified`):
```Rust
         // and only if we are the owner
        if !self.is_owned_by_current_program() {
            return Err(InstructionError::ExternalAccountDataModified);
        }
```

ryoqun commented on sdk/src/transaction_context.rs:
```diff
@@ -215,11 +227,21 @@ impl TransactionContext {
     /// Pops the current InstructionContext
     pub fn pop(&mut self) -> Result<(), InstructionError> {
-        if self.instruction_stack.is_empty() {
-            return Err(InstructionError::CallDepth);
```
hmm, removing this Err code path is okay?

ryoqun commented on sdk/src/transaction_context.rs:
```diff
@@ -215,11 +227,21 @@ impl TransactionContext {
+            instruction_context.instruction_accounts_lamport_sum != instruction_accounts_lamport_sum
+        } else {
+            false
+        };
```
maybe, this comment might help future readers:

```rust
// always `.pop()` even if we detected unblanced instruction to avoid unmatched .push() and .pop()
self.instruction_stack.pop();
```

ryoqun commented on sdk/src/transaction_context.rs:
```diff
@@ -215,11 +227,21 @@ impl TransactionContext {
-        if self.instruction_stack.is_empty() {
-            return Err(InstructionError::CallDepth);
-        }
+        let instruction_context = self.get_current_instruction_context()?;
```
(heh, my review comment is helping already: https://github.com/solana-labs/solana/pull/25899/files#r918481393)

this `?` (and another one below) can potentially create mismatched push() and pop(). is this really safe? could you comment the reason if this is so or adjust the code?

ryoqun commented on sdk/src/transaction_context.rs:
```diff
@@ -215,11 +227,21 @@ impl TransactionContext {
-            return Err(InstructionError::CallDepth);
-        }
+        let instruction_context = self.get_current_instruction_context()?;
+        let detected_unbalanced_instruction = if self.rent.is_some() {
```
nit: little predicate `fn` will make code friender to human:

```rust
let detected_unbalanced_instruction = if self.enable_early_verification() {
}

...

impl TransanctionContext {
   fn enable_early_verification() -> bool {
      self.rent.is_some()
   }
}
```

ryoqun commented:
(CC: @jstarry) i wonder these ic_logger_msg! in retiring code-path (`verify` and `verify_and_update`) might be missed by daap devs or not..:

```rust
                            .map_err(|err| {
                                ic_logger_msg!(
                                    self.log_collector,
                                    "failed to verify account {}: {}",
                                    key,
                                    err
                                );
                                err
                            })?;
```

Especially, They are now unable to see which account was offending (on the other hand, ic was printing the `key`) because of our super terse unit variant `InstructionError`.
maybe should we do the same ic logger thing for the new code path as well?

ryoqun commented on sdk/src/transaction_context.rs:
```diff
@@ -538,6 +611,28 @@ impl<'a> BorrowedAccount<'a> {
 
     /// Assignes the owner of this account (transaction wide)
     pub fn set_owner(&mut self, pubkey: &[u8]) -> Result<(), InstructionError> {
```
(I'm unresolving this review comment again).

i don't think this is addressed, right?

jstarry replied:
The `set_data_length` method is a good example of what I'm proposing. It has the equality check inside.

Lichtso replied:
The check is already there, see line 633: `if self.get_owner().to_bytes() == pubkey`

jstarry replied:
Oh ok, I missed that you added it there. I would like it to be the first check in the method for `set_lamports` and `set_owner` so that the equality check at the call site can be removed. Same applies for `set_data` but with the extra check to avoid expensive comparison if possible.

Lichtso replied:
Already discussed on Slack, but for completeness. The check at the call site is necessary and can not be moved inside the callee, and the check in the callee can not be moved to be first thing to make the check at the call site redundant. That is because ABIv1 and ABIv2 differ in how they deal with faulty accesses. In ABIv1 the runtime always tries to write to all accounts, but only throws errors if the data written is actually different as a workaround. ABIv2 does not do that anymore and always checks the permissions, even if the data written is the same as it already is.

jstarry replied:
Sounds good, thanks for the discussion!

ryoqun commented:
@Lichtso okay, i'm moderately sure this pr contains a _potential_ lof vulnerability or equivalent footgun. :)

the problem is that we stop validating account modifications when entering into the new child CPI for the case of builtin->bpf. I said _potential_ here because there is no code path for the rather unusual cpi usage.

So, a hypothetical attack would be: 1. confuse any native program to mutate victim program's-accounts, also 2. make it to cpi into the victim's program. 3. once successful, the mutation will be committed as the victim program's valid modifications by the runtime at the time of cpi exit verification. Also note those mutated account aren't needed to be _used_ actually. Rarely, deployed programs err out with extra accounts. So, the attacker just need to piggy back these at the end of harmless/easy-to-construct no-op-like victim's instruction.

Previously, this was covered by the `verify()` but not anymore. that verification is now moved inside `get_translated_account()` , which only covers the cpi-entering verification of bpf->{native,bpf}.

At least, i think we should add a specific test case to catch this.

Ideally, I'd suggest we'll introduce a bunch of tests for these CPI-boundaries with comprehensive bpi/builtin combinations.

Lichtso commented:
> I wonder these `ic_logger_msg!` in retiring code-path (`verify` and `verify_and_update`) might be missed

@ryoqun Interesting point. Unfortunately, the `LogCollector` from the program-runtime is not available to the `TransactionContext` in the sdk. However, a better solution would probably be to augment the `InstructionError` enum to allow specifying which transaction account index caused the error.

> The problem is that we stop validating account modifications when entering into the new child CPI for the case of builtin->bpf

You mean the return edge of `InvokeContext::native_invoke()`? In other words the check should be placed around [here](https://github.com/solana-labs/solana/blob/fd5df1cf25999a50c6c2616340f95c2115b585df/programs/bpf_loader/src/syscalls.rs#L2839)?

Lichtso commented on program-runtime/src/invoke_context.rs:
```diff
@@ -860,9 +873,13 @@ impl<'a> InvokeContext<'a> {
                 .feature_set
                 .is_active(&record_instruction_in_transaction_context_push::id())
             {
+                let instruction_accounts_lamport_sum = self
+                    .transaction_context
+                    .instruction_accounts_lamport_sum(instruction_accounts)?;
```
It is already in the `record_instruction_in_transaction_context_push` feature gate and that is [bound to be activated soon](https://github.com/solana-labs/solana/wiki/Feature-Activation-Schedule). The two feature gates should never overlap.

Lichtso replied:
Exactly

jstarry commented on sdk/src/transaction_context.rs:
```diff
@@ -218,8 +235,26 @@ impl TransactionContext {
             return Err(InstructionError::CallDepth);
         }
+        // Verify (before we pop) that the total sum of all lamports in this instruction did not change
+        let detected_an_unbalanced_instruction = if self.rent.is_some() {
```
nit: should use `is_early_verification_of_account_modifications_enabled` here

ryoqun commented:
> > The problem is that we stop validating account modifications when entering into the new child CPI for the case of builtin->bpf
> 
> ... (@Lichtso wrote)

never mind.. another false alarm from me.

I was just forgetting the fact that this pr makes native code bail out immediately as soon as it writes illegally, thanks to online validations inside BorrowedAccount... so obsessed with my own finding (TM). in other words, i mistook that native program is similar to bpf one in this regard.. (i.e. bpf's illegal check is still detected at the return edge.)

> You mean the return edge of `InvokeContext::native_invoke()`? In other words the check should be placed around [here](https://github.com/solana-labs/solana/blob/fd5df1cf25999a50c6c2616340f95c2115b585df/programs/bpf_loader/src/syscalls.rs#L2839)?

that said, to answer your original question, i think the check is already in-place. i.e. the return edge verification ({native, bpf} parent stack <==cpi== bpf child stack) is already enforced, implicitly via `deserialize_parameters()`.

Also, i think there should be no need for explicit return edge verification for ({native, bpf} parent stack <==cpi== native child stack).

ryoqun commented:
Aaand, i think this pr is quite in good shape. I'm about to do the last sanity-check review. thanks for addressing code reviews. :)

ryoqun commented:
> > I wonder these `ic_logger_msg!` in retiring code-path (`verify` and `verify_and_update`) might be missed
> 
> ... a better solution would probably be to augment the `InstructionError` enum to allow specifying which transaction account index caused the error.

yeah. i think so too. could you address that in this pr or follow-up pr?

jstarry reviewed (commented):
It looks like this PR dropped the check where we check for balanced lamports in the caller's accounts before doing a CPI. The issue there is that a caller program could add a lot of lamports to some accounts before calling other programs as long as it balances the lamports before returning

Lichtso commented:
> yeah. i think so too. could you address that in this pr or follow-up pr?

I will do the changes to the `InstructionError` in a follow-up, this PR is already loaded.

Lichtso commented:
@jstarry Huge thanks for finding and solving this! I took the liberty to rename "pre" to "original" and "post" to "current".

He came up with a very elegant solution on how to perform the check to prevent the following attack:

1. Caller adds some high number of lamports to an account they own.
2. Caller calls a callee, which sees that the caller is supposedly rich / loaded. The callee does something unsound.
3. Caller removes the lamports which they added and the invariance over the instruction is maintained.

The trick is to simply check that the invariance of the caller is maintained, not only at its `pop()` but at every CPI `push()` it performs as well.

Lichtso commented on program-runtime/src/invoke_context.rs:
```diff
@@ -1707,7 +1573,13 @@ mod tests {
             );
 
             assert!(result.is_ok());
-            assert_eq!(invoke_context.accounts_data_meter.remaining(), 0);
```
The feature `enable_early_verification_of_account_modifications` is active in these tests, they would not pass otherwise because the `AccountsDataMeter` is no longer updated.

Lichtso commented:
I think the two questions for @jack are minor and can be addressed separately, once he is back. The augmentation of the `InstructionError` will also go in its own PR. In other words, I think this PR is ready to sail.

@ryoqun @jstarry @brooksprumo Mind giving this your final review & approval?

ryoqun commented:
> @ryoqun ... Mind giving this your final review & approval?

sorry, distracted by other thing. I'll do the final review tomorrow.

jstarry commented on sdk/src/transaction_context.rs:
```diff
@@ -667,19 +842,111 @@ impl<'a> BorrowedAccount<'a> {
+
+    /// Returns an error if `get_data_mut` would fail
+    pub fn can_data_be_changed(&self) -> Result<(), InstructionError> {
+        if !self
```
nit: I'm not a huge fan of this early exit behavior. Even after reviewing this PR for awhile I keep forgetting that this is a no-op if the feature isn't active. Can you pull out the feature gate check to the caller?

Lichtso replied:
You have to keep in min that at some point, the feature gate will be through and can be cleaned up. And the cleanup should be simple and its change set minimal, as it is itself a dangerous change. We already had issues with incorrectly removals of feature gate code. So, in that light I would prefer to keep these in the callee, as these have to many call sites which would all be affected.

jstarry replied:
Ok, that's fine with me

jstarry commented on sdk/src/transaction_context.rs:
```diff
@@ -667,19 +842,111 @@ impl<'a> BorrowedAccount<'a> {
+
+    /// Returns an error if `can_data_be_changed`, `set_data` or `set_data_length` would fail
+    pub fn can_data_be_resized(&self, new_length: usize) -> Result<(), InstructionError> {
+        if !self
```
nit: I'd like this check pulled out to the caller as well

Lichtso commented on program-runtime/src/invoke_context.rs:
```diff
@@ -831,26 +837,35 @@ impl<'a> InvokeContext<'a> {
+            if !self
+                .feature_set
+                .is_active(&enable_early_verification_of_account_modifications::id())
+            {
```
Ryo noted that account borrows can fail, introducing a panic source in his first review. So I already changed that to throwing an error a few days ago.

In other words, Point 4 is included in 2 see [`TransactionContext::instruction_account_lamport_sum()`](https://github.com/solana-labs/solana/blob/1da1930c97e4c7d91691392a0c482767bab47945/sdk/src/transaction_context.rs#L322) which throws `InstructionError::AccountBorrowOutstanding`. Might not be obvious enough, but it tries to borrow every unique instruction account, just like `verify_and_update()` did.

jstarry commented on program-runtime/src/invoke_context.rs:
```diff
@@ -861,22 +876,29 @@ impl<'a> InvokeContext<'a> {
+                    .feature_set
+                    .is_active(&enable_early_verification_of_account_modifications::id())
+                {
+                    Ok(())
```
Resolving this since this is covered in the lamports summation

Lichtso commented on programs/bpf_loader/src/syscalls.rs:
```diff
@@ -2663,26 +2664,51 @@ where
+                        .map_err(SyscallError::InstructionError)?;
+                }
+                // The redundant `can_data_be_resized` helps to avoid the expensive data comparison if we can
+                if callee_account
+                    .can_data_be_resized(caller_account.data.len())
+                    .is_ok()
+                    || callee_account.get_data() != caller_account.data
+                {
```
This is a complex one for sure, and it was carefully crafted to minimize the performance overhead (before and after feature activation) and not require a feature gate (stay invariant to the feature activation). For a detailed explaination see: https://github.com/solana-labs/solana/pull/25899#discussion_r897409117.

> the name of the `can_data_be_resized` method doesn't fully indicate that `set_data` will succeed.

The comment for `can_data_be_resized` states that, but I agree that the naming could be better. Suggestions?

> We should skip the data comparison if the data length changed in an invalid way

Is already the case, if the data length changed in an invalid way `can_data_be_resized` errors and the data comparison never happens.

> We should reuse the errors from check a) if the data was changed rather than recomputing that error inside set_data.

In ABIv1, because of the workaround which was necessary to make it work with `PreAccount`, all accounts are always written back. So readonly accounts produce incorrect errors and we need to ignore not (re)use them.

jstarry replied:
> The comment for can_data_be_resized states that, but I agree that the naming could be better. Suggestions?

I think your names are fine, it's just that I don't like that `can_data_be_resized` includes the behavior of `can_data_be_changed` as well. I think that the redundant check should check if both `can_data_be_changed` and `can_data_be_resized` return ok before skipping the data comparison.

> Is already the case, if the data length changed in an invalid way can_data_be_resized errors and the data comparison never happens.

When `can_data_be_resized` errors, we always continue to line 2676 which does a data comparison

> In ABIv1, because of the workaround which was necessary to make it work with PreAccount, all accounts are always written back. So readonly accounts produce incorrect errors and we need to ignore not (re)use them.

I understand that. I'm talking about something slightly different. In case 2 above, we know that the data change will fail if the data comparison returns not-equal, so we should reuse that error that we already computed by calling the "redundant" `can_data_be_resized` method.

Lichtso replied:
I split `can_data_be_resized` and `can_data_be_changed` to be independent of one another. And I moved `can_data_be_resized()?` out of the `if` so that it can lead to early termination.

jstarry replied:
The new code only detects invalid resizes for unowned accounts. If an owned executable or read-only account is resized, we will do a full data comparison even though checking the length is sufficient. Might be good to fix that by adding more checks in `can_data_be_resized`?

And for the record, here's the code that I think would be more readable:
```rust
        // The redundant check helps to avoid the expensive data comparison if we can
        match borrowed_account.can_data_be_changed() {
            Ok(()) => borrowed_account.set_data(&data)?,
            Err(err) => {
                let current_data = borrowed_account.get_data();
                if current_data.len() != data.len() || current_data != data {
                    return Err(err);
                }
            }
        }
```

jstarry replied:
@Lichtso I see you resolved this, anything to comment about the edge cases I mentioned above for avoiding data comparison?

Lichtso replied:
Added your `match` based code. But `can_data_be_resized()` needs to stay and come before `can_data_be_changed()` otherwise the error priority of `InstructionError::AccountDataSizeChanged` would change.

jstarry commented on sdk/src/transaction_context.rs:
```diff
@@ -667,19 +848,111 @@ impl<'a> BorrowedAccount<'a> {
+        Ok(())
+    }
+
+    /// Returns an error if `can_data_be_changed`, `set_data` or `set_data_length` would fail
```
nit: I think this comment and the comment for `can_data_be_changed` needs to be updated

jstarry reviewed (dismissed):
At this point, I feel pretty comfortable with the soundness of the PR changes. I've left follow up comments on suggested cleanup but if you disagree, I'm ok with this being merged as is. Thanks for the rigor and quickness in addressing feedback!

ryoqun commented on sdk/src/transaction_context.rs:
```diff
@@ -251,6 +304,40 @@ impl TransactionContext {
+            instruction_accounts_lamport_sum = (self
+                .get_account_at_index(instruction_account.index_in_transaction)?
+                .try_borrow()
+                .map_err(|_| InstructionError::AccountBorrowOutstanding)?
```
the `AccountBorrowOutstanding` check starts to be ommitted for program_accounts it seems: https://github.com/solana-labs/solana/blob/3e9bd6170c9d78db850c9c7c7bcd832b4ce311bd/program-runtime/src/invoke_context.rs#L476-L481

Lichtso replied:
Good finding, seems like this was only enforced in `verify()` (at transaction instruction level), not in CPI (`verify_and_update()`). Nonetheless, I included it to happen in every instruction `pop()`.

Lichtso commented:
I think I addressed all review comments, CI is passing, no merge conflicts ...
@jstarry @ryoqun @brooksprumo Can you once more approve so we can finally merge this?

ryoqun commented on sdk/src/transaction_context.rs:
```diff
@@ -218,8 +251,34 @@ impl TransactionContext {
+                .and_then(|instruction_context| {
+                    // Verify all executable accounts have no outstanding refs
+                    for account_index in instruction_context.program_accounts.iter() {
+                        self.get_account_at_index(*account_index)?
```
hey, i know you're itching to merge this great pr... but.

this `?` can create unbalanced `self.instruction_stack`...

ryoqun replied:
oh, never mind...

 this is actually is in a `.and_then()` closure?

ryoqun reviewed (approved):
lgtm. master piece pr. :)
